oml/metrics/embeddings.py

```diff
@@ -147,12 +169,104 @@ def compute_metrics(self) -> TMetricsDict_ByLabels:  # type: ignore
                     distances=self.distance_matrix[mask],  # type: ignore
                     mask_gt=self.mask_gt[mask],  # type: ignore
                     mask_to_ignore=self.mask_to_ignore[mask],  # type: ignore
+                    reduce=False,
                     **args,  # type: ignore
                 )
 
-        self.metrics = metrics  # type: ignore
+        self.metrics_unreduced = metrics
+        self.metrics = reduce_metrics(metrics)  # type: ignore
+
+        return self.metrics
+
+    def get_worst_queries_ids(self, metric_name: str, topk: int) -> List[int]:
+        metric_values = flatten_dict(self.metrics_unreduced)[metric_name]  # type: ignore
+        return np.argsort(metric_values)[:topk].tolist()
+
+    def get_plot_for_worst_queries(
+        self, metric_name: str, topk_queries: int, topk_instances: int, verbose: bool = False
+    ) -> plt.Figure:
+        query_ids = self.get_worst_queries_ids(metric_name=metric_name, topk=topk_queries)
+        return self.get_plot_for_queries(query_ids=query_ids, top_k=topk_instances, verbose=verbose)
+
+    def get_plot_for_queries(
+        self,
+        query_ids: List[int],
+        top_k: int,
+        verbose: bool = True,
+    ) -> plt.Figure:
+        """
+        Visualize the predictions for the query with the indicies <query_ids>.
+
+        Args:
+            query_ids: Index of the query
+            top_k: Amount of the predictions to show
+            verbose: wether to show image paths or not
+
+        """
+        assert self.metrics is not None, "We are not ready to plot, because metrics were not calculated yet."
+
+        dist_matrix_with_inf = torch.clone(self.distance_matrix)
+        dist_matrix_with_inf[self.mask_to_ignore] = float("inf")
+
+        is_query = self.acc.storage[self.is_query_key]
+        is_gallery = self.acc.storage[self.is_gallery_key]
+
+        query_paths = np.array(self.acc.storage[PATHS_KEY])[is_query]
+        gallery_paths = np.array(self.acc.storage[PATHS_KEY])[is_gallery]
+
+        fake_coord = np.array([float("nan")] * len(is_query))
+        bboxes = list(
+            zip(
+                self.acc.storage.get(X1_KEY, fake_coord),
+                self.acc.storage.get(Y1_KEY, fake_coord),
+                self.acc.storage.get(X2_KEY, fake_coord),
+                self.acc.storage.get(Y2_KEY, fake_coord),
+            )
+        )
+
+        query_bboxes = torch.tensor(bboxes)[is_query]
+        gallery_bboxes = torch.tensor(bboxes)[is_gallery]
+
+        fig = plt.figure(figsize=(30, 30 / (top_k + 2 + 1) * len(query_ids)))
+        for j, query_idx in enumerate(query_ids):
+            ids = torch.argsort(dist_matrix_with_inf[query_idx])[:top_k]
+
+            n_gt = self.mask_gt[query_idx].sum()  # type: ignore
```

Comment:
Wrong behaviour cases for cases with ReID

function `apply_mask_to_ignore` allow you to exclude some of GT, so use it